rest/endpoint/v1/api.go

```diff
@@ -0,0 +1,874 @@
+	tableName := s.params(r, tableParam)
+	rowIdentifier := s.params(r, "rowIdentifier")
+	user := auth.ContextUserOrRole(r.Context())
+
+	// grab table and extract primary key col names
+	columns, values, err := primaryKeyValues(rowIdentifier, keyspaceName, tableName, s.dbClient)
+	if err != nil {
+		msg := "unable to get primary key"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "table", tableName, "error", err)
+
+		switch err.(type) {
+		case *e.InternalError:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusInternalServerError)
+			return
+		default:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusBadRequest)
+			return
+		}
+	}
+
+	_, err = s.dbClient.Delete(&db.DeleteInfo{
+		Keyspace:    keyspaceName,
+		Table:       tableName,
+		Columns:     columns,
+		QueryParams: values,
+	}, newDbOptions(user))
+
+	if err != nil {
+		msg := "unable to execute delete query"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "table", tableName, "error", err)
+		RespondWithError(w, errors.New(msg), http.StatusInternalServerError)
+		return
+	}
+
+	RespondJSONObjectWithCode(w, http.StatusNoContent, nil)
+}
+
+func (s *routeList) GetTables(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	keyspaceName := s.params(r, keyspaceParam)
+	user := auth.ContextUserOrRole(r.Context())
+
+	tables, err := s.dbClient.DescribeTables(keyspaceName, user)
+	if err != nil {
+		msg := "unable to describe tables"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "error", err)
+
+		switch err.(type) {
+		case *e.NotFoundError:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusNotFound)
+			return
+		default:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	RespondJSONObjectWithCode(w, http.StatusOK, tables)
+}
+
+func (s *routeList) GetTable(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	keyspaceName := s.params(r, keyspaceParam)
+	tableName := s.params(r, tableParam)
+	user := auth.ContextUserOrRole(r.Context())
+
+	table, err := s.dbClient.DescribeTable(keyspaceName, tableName, user)
+	if err != nil {
+		msg := "unable to describe table"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "table", tableName, "error", err)
+
+		switch err.(type) {
+		case *e.NotFoundError:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusNotFound)
+			return
+		default:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	RespondJSONObjectWithCode(w, http.StatusOK, tableMetadataToTable(table))
+}
+
+func (s *routeList) AddTable(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	keyspaceName := s.params(r, keyspaceParam)
+	user := auth.ContextUserOrRole(r.Context())
+
+	var tableAdd m.TableAdd
+	if err := parseAndValidatePayload(&tableAdd, r); err != nil {
+		msg := "unable to parse payload"
```

Comment:
This type of opaque error is not a good user experience. For example, I was missing the primary key and all I got back was "unable to parse payload".  For all I knew it was invalid json.

Reply:
Created: https://github.com/datastax/cassandra-data-apis/issues/129